File: CALISTA-MATLAB/subfunctions/get_3D_cell_prob.m
function [cell_prob_3D]=get_3D_cell_prob(Results,Parameters,DATA)


mRNA_all=DATA.totDATA;
final_groups=Results.final_groups;
log_p_mat_ma=Parameters.logP;
p_mat_ma=Parameters.P;
nvars=DATA.nvars;
n_genes=DATA.numGENES;

% mRNA counts of lookup table
[max_mRNA_counts, ~]=size(log_p_mat_ma);
% zero is not included, therefore -1
max_mRNA_counts=max_mRNA_counts-1;
% define array
num_cells_mRNA=zeros(max_mRNA_counts+1,n_genes);

% sort clusters
[sorted_as,sortedIDX]=sort(final_groups);
% sort mRNA based on clusters
data_sorted=mRNA_all(sortedIDX,:);
% calculate unique clusters
[clusters,lastIDX]=unique(sorted_as,'last');
% calculate number of clusters
num_clusters=length(clusters);
% bounds of each clusters
bounds=[1 lastIDX(1:end-1)'+1; lastIDX']';
% invert the probability matrix
X=log_p_mat_ma';
% define array
opt_idx_clusters=zeros(num_clusters,n_genes);
%loop over clusters
for clust=1:num_clusters
    %  clust
    cells_in_each_cluster=data_sorted(bounds(clust,1):bounds(clust,2),:);
    % if cluster has only one entry
    if bounds(clust,1)==bounds(clust,2)
        idx = sub2ind(size(num_cells_mRNA),cells_in_each_cluster+1, 1:n_genes);
        num_cells_mRNA(idx)=1;
    else
        % number of cells with mRNA count
        num_cells_mRNA(:,:)=hist(cells_in_each_cluster,0:1:max_mRNA_counts);
    end
    % Matrixmultiplication
    Y=sparse(num_cells_mRNA);
    Y=Y/sum(Y(:,1));
    Z=X*Y;
    % get the maximum for each gene
    [~,idx_max_L]=max(Z);
    opt_idx_clusters(clust,:)=idx_max_L;
end

% loop over all cells
for i=1:nvars
    
    % loop over cluster
    for clust=1:num_clusters
        % loop over genes
        for j=1:n_genes
            % calculate the lop prob of gene in cell belonging
            % to cluster
            opt_param_each_gene=p_mat_ma(:,opt_idx_clusters(clust,j));
            % sum all genes together
            cell_prob_3D(i,clust,j)=opt_param_each_gene(mRNA_all(i,j)+1);
        end
    end
end

            
            

File: CALISTA-R/calista/R/get_3D_cell_prob.R
#' A calista Function
#'
#' @param Results,Parameters,DATA
#' @keywords calista
#' @export
#' @examples
#' get_3D_cell_prob()






get_3D_cell_prob<-function(Results,Parameters,DATA){

  mRNA_all=DATA$totDATA
  final_groups=Results$final_groups
  log_p_mat_ma=Parameters$Parameters[[3]]
  p_mat_ma=Parameters$Parameters[[2]]
  nvars=DATA$nvars
  n_genes=DATA$numGENES
  #mRNA counts of lookup table
  max_mRNA_counts=nrow(log_p_mat_ma)
  #zero is not included, therefore -1
  max_mRNA_counts=max_mRNA_counts-1
  #define array
  num_cells_mRNA=matrix(0,max_mRNA_counts+1,n_genes)
  #sort cluster
  sorted_as=sort(final_groups)
  sortedIDX=order(final_groups)
  #sort mRNA based on clusters
  data_sorted=mRNA_all[sortedIDX,]
  #calculate unique clusters
  Clusters=unique(sorted_as)
  lastIDX=length(sorted_as)-match(Clusters,rev(sorted_as))+1
  #calculate number of clusters
  num_clusers=length(Clusters)
  #bouds of each Clusters
  bounds=as.matrix(data.frame(c(1,lastIDX[1:length(lastIDX)-1]+1),lastIDX))
  #invert the prabalitity matrix
  X=t(log_p_mat_ma)
  #define array
  opt_idx_clusters=matrix(0,num_clusers,n_genes)
  #loop over each cluster
  for (clust in 1:num_clusers){
    cells_in_each_cluster=data_sorted[bounds[clust,1]:bounds[clust,2],]
    if(bounds[clust,1] == bounds[clust,2]){
      idx=((1:n_genes)-1)*nrow(num_cells_mRNA)+cells_in_each_cluster+1
      num_cells_mRNA[idx]=1
    }else{
      num_cells_mRNA=apply(cells_in_each_cluster,2,function(x){
          hist(x,seq(-0.5,max_mRNA_counts+0.5,1),plot = FALSE)[[2]]
        })
    }
    Y=Matrix(num_cells_mRNA,sparse = TRUE)
    Y=Y /sum(Y[,1])
    Z=X %*% Y
    idx_max_L=apply(Z,2,function(x){
      which(x==max(x))[1]
    })
    opt_idx_clusters[clust,]=idx_max_L
  }
  cell_prob_3D=array(0,c(nvars,num_clusers,n_genes))
  #loop over all cells
  for(i in 1:nvars){
    for(clust in 1:num_clusers){
      for(j in 1:n_genes){
        opt_param_each_gene=p_mat_ma[,opt_idx_clusters[clust,j]]
        cell_prob_3D[i,clust,j]=opt_param_each_gene[mRNA_all[i,j]+1]
      }
    }
  }
  return(cell_prob_3D)

}
